FSDP/T5_training.py

```diff
@@ -0,0 +1,346 @@
+    and dist.is_nccl_available()
+    and nccl.version() >= (2, 10)
+    )
+    
+    if bf16_ready:
+        mp_policy = bfSixteen
+    else:
+        mp_policy = None # defaults to fp32
+        
+    model = FSDP(model,
+        auto_wrap_policy=t5_auto_wrap_policy,
+        mixed_precision=mp_policy,
+        #sharding_strategy=sharding_strategy,
+        device_id=torch.cuda.current_device())
+
+    optimizer = optim.AdamW(model.parameters(), lr=args.lr)
+
+    scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
+    best_val_loss = float("inf")
+    curr_val_loss = float("inf")
+    file_save_name = "T5-model-"
+
+    if rank == 0:
+        time_of_run = get_date_of_run()
+        dur = []
+        train_acc_tracking = []
+        val_acc_tracking = []
+        training_start_time = time.time()
+
+    if rank == 0 and args.track_memory:
+        mem_alloc_tracker = []
+        mem_reserved_tracker = []
+
+    for epoch in range(1, args.epochs + 1):
+        t0 = time.time()
+        train_accuracy = train(args, model, rank, world_size, train_loader, optimizer, epoch, sampler=sampler1)
+        if args.run_validation:
+            curr_val_loss = validation(model, rank, world_size, val_loader)
+        scheduler.step()
+        
+        if rank == 0:
+
+            print(f"--> epoch {epoch} completed...entering save and stats zone")
+
+            dur.append(time.time() - t0)
+            train_acc_tracking.append(train_accuracy.item())
+
+            if args.run_validation:
+                val_acc_tracking.append(curr_val_loss.item())
+
+            if args.track_memory:
+                mem_alloc_tracker.append(
+                    format_metrics_to_gb(torch.cuda.memory_allocated())
+                )
+                mem_reserved_tracker.append(
+                    format_metrics_to_gb(torch.cuda.memory_reserved())
+                )
+            print(f"completed save and stats zone...")
+        
+
+    #init_end_event.record()
+        
+    #if rank == 0:
+        #print(f"Cuda event elapsed time: {init_start_event.elapsed_time(init_end_event) / 1000}sec")
+        # print(f"{model}")
+
+        if args.save_model and curr_val_loss < best_val_loss:
+            
+            # save
+            if rank == 0:
+                print(f"--> entering save model state")
+            
+            save_policy = FullStateDictConfig(offload_to_cpu=True, rank0_only=True)
+            with FSDP.state_dict_type(
+                model, StateDictType.FULL_STATE_DICT, save_policy
+            ):
+                cpu_state = model.state_dict()
+            #print(f"saving process: rank {rank}  done w state_dict")
+           
+
+            if rank == 0:
+                print(f"--> saving model ...")
+                currEpoch = (
+                    "-" + str(epoch) + "-" + str(round(curr_val_loss.item(), 4)) + ".pt"
+                )
+                print(f"--> attempting to save model prefix {currEpoch}")
```

Comment:
nit: saving could be its own function

Reply:
agree - it's done this way in a different fsdp repo I have.

Reply:
updated